### ISSUE TITLE
Support for arm (rpi)

### ISSUE THREAD

**Issue by RaymondMouthaan** [vfarcic/docker-flow-proxy#449]
Hello @vfarcic 

Currently i am using a 5 rpi docker swarm cluster with HAProxy  with several backends. Now i came across your project and very well documented site. I am really interested to use this project for my swarm. Unfortunately I couldn't find the arm docker images for it.

Are there any plans to build the arm images as well? If needed it like to work together to make it available too for arm. 

After searching i found several projects to support arm, but they are quite out-dated or undocumented.

- docker-hub: taylorsmcclure/docker-flow-arm (more than a year old) -> https://github.com/mastermindg/docker-flow-proxy

- docker-hub: monsonnl/docker-flow-proxy (about 8 months old) -> https://github.com/monsonnl/docker-flow-proxy

- docker-hub: dsbaha/docker-flow-proxy (about 3 months old), but no GitHub available. 

Deploying the example docker-compose-stack.yml file with some modifications to use the dsbaha/docker-flow-proxy and dsbaha/docker-flow-swarm-listener images, seems to run fine.

```
version: "3"

services:

  proxy:
    image: dsbaha/docker-flow-proxy
    ports:
      - 9011:80
      - 9443:443
    networks:
      - proxy
    environment:
      - LISTENER_ADDRESS=swarm-listener
      - MODE=swarm
    deploy:
      replicas: 2

  swarm-listener:
    image: dsbaha/docker-flow-swarm-listener
    networks:
      - proxy
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    environment:
      - DF_NOTIFY_CREATE_SERVICE_URL=http://proxy:8080/v1/docker-flow-proxy/reconfigure
      - DF_NOTIFY_REMOVE_SERVICE_URL=http://proxy:8080/v1/docker-flow-proxy/remove
    deploy:
      placement:
        constraints: [node.role == manager]

networks:
  proxy:
    external: true
```

![screen shot 2018-03-25 at 13 24 24](https://user-images.githubusercontent.com/25717738/37874459-e00eeadc-302f-11e8-82d6-429aba17d3ea.png)

Best regards,
Raymond

**Comment by vfarcic** [comment]
I'm not very familiar with ARM. Can you send a PR with, let's say, Dockerfile.rpi? If you can, I can add it to our Jenkins processes so that an ARM release is done with every commit.

**Comment by RaymondMouthaan** [comment]
Victor, thanks for the reply.

What I will try do is to build a docker image for your project using Travis as I already have some build some other images (see my github repo) this way. When it works I'll let you know and maybe you can us it as an example for your Jenkins process, so it gets part of your project instead of one of my own.

I saw there is a docker-flow-proxy_linux_arm release available, I maybe could use that or maybe build it from source?

What is required for dfp? Like Go version, OS, haproxy?

**Comment by vfarcic** [comment]
The release process already creates ARM binaries. The only thing needed is to create a Dockerfile that will be used to build ARM images.

Whatever works best for you. You can do it with Travis it takes is easier. From my side, if you add Dockerfile to the repo, I think I can take care of the process.

I haven't tested DFP with older Go versions. I guess it should work with any. As for HAProxy, it's best if it's the same as the one used in https://github.com/vfarcic/docker-flow-proxy/blob/master/Dockerfile (1.8.3). As for OS (assuming that you're referring to the one used in Dockerfile) I use Alpine but I'm not sure which one is preferable for ARM.

**Comment by RaymondMouthaan** [comment]
Hello @vfarcic ,

I've managed to build **docker-flow-proxy** and **docker-flow-swarm-listener** for both arm and amd64 based upon the provided binaries. 

The projects can be found at [docker-flow-proxy](https://github.com/RaymondMouthaan/docker-flow-proxy) and [docker-flow-swarm-listener](https://github.com/RaymondMouthaan/docker-flow-swarm-listener).

Just to summarise ...
**docker-flow-proxy**
Dockerfile.linux-arm uses the latest **arm32v7/haproxy** as base image, which has Debian lite as os. It copies **qemu-arm-static** (which is provided by the .travis.yml) This file is needed to build arm images on x86_64. Using curl, it downloads the DFP binary from your repo and places it in /usr/local/bin. All other lines are pretty much the same as your Dockerfile.
Same goes for Dockerfile.linux-amd64, except for **qemu-arm-static**, which is not needed. 
The travis build job builds both the linux-arm and linux-amd64 images and pushes it to docker hub and uses the manifest-tool to create and push manifest-list as well. 

**docker-flow-swarm-listener**
Dockerfile.linux-arm uses the latest **arm32v6/alpine** as base image, which has Alpine as os. It copies **qemu-arm-static** (which is provided by the .travis.yml) This file is needed to build arm images on x86_64. Using wget, it downloads the DFSL binary from your repo and places it in /usr/local/bin. All other lines are pretty much the same as your Dockerfile.
Dockerfile.linux-amd64 is the same except for qemu.

Running Docker Flow Proxy on my Synology NAS (x86_64) the logging shows these lines, so i assume  it runs fine.
![screen shot 2018-03-27 at 08 36 46](https://user-images.githubusercontent.com/25717738/37950756-0fd9b0d8-319a-11e8-8c49-af4c21870c79.png)

Same log lines are shown when running on my raspberry pi docker node.

As for docker-proxy-swarm-listener on the Synology, it gives the error below, but this is most probably because i did not started with the correct settings, but it looks promising.
![screen shot 2018-03-27 at 08 43 39](https://user-images.githubusercontent.com/25717738/37950976-02a2f77a-319b-11e8-8d59-e91aaebd3daf.png)
I'll continue working on that.

As for integration with your Jenkins process. I guess if you are able to use qemu to build the arm images on x86_64 systems (assuming you have), you could use the Dockerfiles to build the images and using manifest-tool to create and push manifest-list to docker hub.

Personally, I prefer to keep everything in one place, let me know what you think :-)

**Comment by vfarcic** [comment]
Sorry for not getting back to you earlier. Too much work...

This looks great. Please send a PR with the new Dockerfile* (whenever you think they're ready) and I'll add it to Jenkinsfile. From there on, ARM image will be available with each release :)

**Comment by RaymondMouthaan** [comment]
That's oke, I have been busy myself too.

Later today I'll do the PR as you requested 👍🏼. Also I will have a look at your docker proxy monitor project to have a dockerfile for arm too 💪.

**Comment by RaymondMouthaan** [comment]
Btw using qemu isn't an issue for you from Jenkins?

**Comment by vfarcic** [comment]
It's available from the [release 18.03.31-121](https://github.com/vfarcic/docker-flow-proxy/releases/tag/18.03.31-121).
